pkg/route/validation/validation.go

```diff
@@ -354,6 +308,62 @@ func validateTLS(route *routev1.Route, fldPath *field.Path) field.ErrorList {
 	return result
 }
 
+// validateTLSExternalCertificate tests different pre-conditions required for
+// using externalCertificate. Called by validateTLS.
+func validateTLSExternalCertificate(ctx context.Context, route *routev1.Route, fldPath *field.Path, sarc routecommon.SubjectAccessReviewCreator, secretsGetter corev1client.SecretsGetter) field.ErrorList {
+	tls := route.Spec.TLS
+
+	// user must have create and update permission on the custom-host sub-resource.
+	errs := routecommon.CheckRouteCustomHostSAR(ctx, fldPath, sarc)
+
+	// The router serviceaccount must have permission to get/list/watch the referenced secret.
+	// The role and rolebinding to provide this access must be provided by the user.
+	if err := authorizationutil.Authorize(sarc, &user.DefaultInfo{Name: routerServiceAccount},
+		&authorizationv1.ResourceAttributes{
+			Namespace: route.Namespace,
+			Verb:      "get",
+			Resource:  "secrets",
+			Name:      tls.ExternalCertificate.Name,
+		}); err != nil {
+		errs = append(errs, field.Forbidden(fldPath, "router serviceaccount does not have permission to get this secret"))
+	}
+
+	if err := authorizationutil.Authorize(sarc, &user.DefaultInfo{Name: routerServiceAccount},
+		&authorizationv1.ResourceAttributes{
+			Namespace: route.Namespace,
+			Verb:      "watch",
+			Resource:  "secrets",
+			Name:      tls.ExternalCertificate.Name,
+		}); err != nil {
+		errs = append(errs, field.Forbidden(fldPath, "router serviceaccount does not have permission to watch this secret"))
+	}
+
+	if err := authorizationutil.Authorize(sarc, &user.DefaultInfo{Name: routerServiceAccount},
+		&authorizationv1.ResourceAttributes{
+			Namespace: route.Namespace,
+			Verb:      "list",
+			Resource:  "secrets",
+			Name:      tls.ExternalCertificate.Name,
+		}); err != nil {
+		errs = append(errs, field.Forbidden(fldPath, "router serviceaccount does not have permission to list this secret"))
+	}
+
+	// The secret should be in the same namespace as that of the route.
+	secret, err := secretsGetter.Secrets(route.Namespace).Get(ctx, tls.ExternalCertificate.Name, metav1.GetOptions{})
+	if err != nil && apierrors.IsNotFound(err) {
+		return append(errs, field.NotFound(fldPath, err))
+	} else if err != nil {
+		return append(errs, field.InternalError(fldPath, err))
+	}
+
+	// The secret should be of type kubernetes.io/tls
+	if secret.Type != corev1.SecretTypeTLS {
+		errs = append(errs, field.Invalid(fldPath, tls.ExternalCertificate.Name, "secret of type 'kubernetes/tls' required"))
```

Comment:
That should be "kubernetes.io/tls", or better, just build the error message using the named const:
```suggestion
		errs = append(errs, field.Invalid(fldPath, tls.ExternalCertificate.Name, fmt.Sprintf("secret of type %q required", corev1.SecretTypeTLS)))
```

Reply:
Thanks, I've updated the error message.